packages/workflow/src/workflow.ts

```diff
@@ -763,39 +762,27 @@ export class Workflow {
 	}
 
 	getParentMainInputNode(node: INode): INode {
```

Comment:
I prefer the more declarative (functional programming) to the iterative way and if it doesn't decrease performance I'd go with this
```suggestion
	getParentMainInputNode(node: INode): INode {
  if (!node) return node;

  const nodeConnections = this.connectionsBySourceNode[node.name];
  if (!nodeConnections) return node;

  const nonMainConnectionTypes = Object.keys(nodeConnections).filter(
    (type) => type !== NodeConnectionTypes.Main,
  );

  const parentNode = nonMainConnectionTypes
    .flatMap((connectionType) => nodeConnections[connectionType] || [])
    .flatMap((connectionGroup) => connectionGroup || [])
    .map((connection) => {
      if (connection?.node) {
        const returnNode = this.getNode(connection.node);
        if (!returnNode) {
          throw new ApplicationError(`Node "${connection.node}" not found`);
        }
        return this.getParentMainInputNode(returnNode);
      }
      return null;
    })
    .find((result) => result !== null);

  return parentNode || node;
}
```